src/operator/tensor/cast_storage-inl.h

```diff
@@ -47,30 +47,34 @@ struct MarkRspRowIdx {
  * CPU implementation of casting a dns tensor to rsp type.
  */
 inline void CastStorageDnsRspImpl(mshadow::Stream<cpu>* s, const TBlob& dns, NDArray* rsp) {
+  using namespace rowsparse;
+  using namespace mshadow;
   CHECK(rsp != nullptr);
   CHECK_EQ(rsp->storage_type(), kRowSparseStorage);
   CHECK_EQ(dns.shape_, rsp->shape());
   MSHADOW_TYPE_SWITCH(dns.type_flag_, DType, {  // data type
-    MSHADOW_IDX_TYPE_SWITCH(rsp->aux_type(rowsparse::kIdx), RType, {  // row idx type
+    MSHADOW_IDX_TYPE_SWITCH(rsp->aux_type(kIdx), RType, {  // row idx type
       const index_t num_rows = dns.shape_[0];
-      const index_t num_cols = dns.shape_[1];
-      rsp->CheckAndAllocAuxData(rowsparse::kIdx, mshadow::Shape1(num_rows));
-      TBlob row_idx_blob = rsp->aux_data(rowsparse::kIdx);
+      const index_t row_length = dns.shape_.ProdShape(1, dns.shape_.ndim());
+      rsp->CheckAndAllocAuxData(kIdx, Shape1(num_rows));
+      TBlob row_idx_blob = rsp->aux_data(kIdx);
       RType* row_idx = row_idx_blob.dptr<RType>();
       mxnet_op::Kernel<MarkRspRowIdx, cpu>::Launch(s, num_rows, row_idx,
-          dns.dptr<DType>(), num_cols);
+          dns.dptr<DType>(), row_length);
       index_t nnr = 0;
-      nnr = mxnet::common::ParallelAccumulate(row_idx, num_rows, nnr);
-      rsp->set_aux_shape(rowsparse::kIdx, mshadow::Shape1(nnr));
+      nnr = common::ParallelAccumulate(row_idx, num_rows, nnr);
+      rsp->set_aux_shape(kIdx, Shape1(nnr));
       if (0 == nnr) return;
-      rsp->CheckAndAllocData(mshadow::Shape2(nnr, num_cols));
-      mshadow::Tensor<cpu, 2, DType> dns_data = dns.FlatTo2D<cpu, DType>(s);
-      mshadow::Tensor<cpu, 2, DType> rsp_data = rsp->data().FlatTo2D<cpu, DType>(s);
+      auto storage_shape = dns.shape_;
+      storage_shape[0] = nnr;
+      rsp->CheckAndAllocData(storage_shape);
+      auto dns_data = dns.get_with_shape<cpu, 2, DType>(Shape2(num_rows, row_length), s);
+      auto rsp_data = rsp->data().get_with_shape<cpu, 2, DType>(Shape2(nnr, row_length), s);
```

Comment:
changing `FlatTo2D` involves a lot more efforts since it's used many places else than sparse tensor.
@piiswrong could you confirm that https://github.com/dmlc/mshadow/blob/a7569f1311f98f952eb5a068f97d5cda4526e532/mshadow/tensor.h#L111  the `FlatTo2D ` function is a bug?